### ISSUE TITLE
new lint: let_underscore_var

### ISSUE THREAD

**Issue by xxchan** [rust-lang/rust-clippy#9678]
### What it does

Unlike `let _ = <expr>` (it's a drop), `let _ = <variable>` is a no-op. I believe it's useless.

Also the discription of [let_underscore_drop](https://rust-lang.github.io/rust-clippy/master/#let_underscore_drop) maybe inprecise becasue it ignored this case.

### Lint Name

let_underscore_var

### Category

correctness, suspicious

### Example

```rust
let _ = x;
```
should warn: this is a no-op. Use `let _x = x;` to move `x` but extend it's lifetime until the end of the block, or use `mem::drop(x)` to drop `x`;

**Comment by blyxyas** [comment]
The lint warned about Clippy internals tools, this was a warning sign about this lint because Rust's official code is well written, it turns out that `let _ = <var>` is used.

As explained in [the RFC](https://rust-lang.github.io/rfcs/2526-const-wildcard.html):

> ## Motivation
> The ability to ensure that code type checks while discarding the result is useful, especially in custom derives. For exampe, the following code will not compile if the type `MyType` doesn't implement the trait `MyTrait`:
> ```
> const _FOO: () = {
>    use std::marker::PhantomData;
>    struct ImplementsMyTrait<T: MyTrait>(PhantomData<T>);
>    let _ = ImplementsMyTrait::<MyType>(PhantomData); // type checking error if MyType: !MyTrait
>    ()
> };
> ```

I would suggest to close this issue, as the problem the lints talks about isn't really a problem as it is a feature.
I just wasted a whole day :/

**Comment by xxchan** [comment]
I'm glad to know that `let _ = <var>` has some use cases. But I still think the rationale of the lint makes sense. Can you show some relevant code about how the feature is used?

**Comment by xxchan** [comment]
IIUC `write!` should expand to a function call and returns a `Result`, so it's a `let _ = <expr>` instead of `let _ = <var>`? 🤔
